Bugfix/biagas/consistent zlib usage

biagas opened the pull request:
### Description

Related to #2679
Make visit's usage of zlib consistent throughout source and third party libs: either use the one
built by build_visit or use system zlib.

Removed use of vtk's zlib from within VisIt.
Removed zlib from visit's third_party_builtin.
Modified bv_zlib.sh to export it's paths for other modules to use.
Modified bv_vtk.sh to not build its own zlib but to use either the zlib built by build_visit or system zlib.
Modified bv_xdmf.sh to point to build_visit's zlib or system zlib.
Modified bv_hdf4.sh to find system zlib if --zlib not specified.

For our usage on lc (where we normally build zlib with build_visit), will require vtk and xdmf to be rebuilt).

### Type of change

Please select one below (*Please click check boxes AFTER submitting ticket*)

- [ ] Bug fix
- [ ] New feature
- [ ] New Documentation
- [ ] Other (please describe below)

### How Has This Been Tested?
On pascal:
I ran build_visit with --zlib to build all tp libs. I then compiled visit against these libs and ran the test suite. Everything passes.

I ran build_visit without --zlib, so it would use system zlib, and built all tp libs.  I then compiled visit against these libs and ran the test suite. Everything passes.

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the release notes
- [ ] I have made corresponding changes to the documentation
- [ ] I have added debugging support to my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] I have added any new baselines to the repo
- [X] I have assigned reviewers (see [VisIt's PR procedures](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers) for more information).

biagas committed:
Initial pass at zlib consistency.
Use system or build_visit version.
Make thirdparty lib's, including VTK, do the same.
Remove zlib from third_party_builtin.

biagas committed:
Remove vtk_zlib.h inclusion. Fix windows config-site file.

biagas committed:
Use cmake command line to find system zlib if --zlib not requested.

biagas committed:
Merge branch 'bugfix/biagas/consistent_zlib_usage' of https://github.com/visit-dav/visit into bugfix/biagas/consistent_zlib_usage

biagas committed:
Remove vtk_zlib.h in favor of zlib.h

biagas committed:
Use zlib.h instead of vtk_zlib.h.

biagas committed:
Merge branch 'develop' into bugfix/biagas/consistent_zlib_usage

biagas committed:
Add avtLAMMPSFileFormat.C to xml file for auto-regeneration of CMakeLists.txt

markcmiller86 commented:
Haven't looked in detail yet but both `scalable.py` and, more importantly, `sclable_c.py` (which tests compression between engine and viewer) are in skip list. Can you at least run `scalable_c.py` explicitly and see if it a) doesn't crash and b) renders images that kinda sorta look right?

markcmiller86 commented on src/config-site/windows.cmake:
```diff
@@ -135,7 +135,7 @@ VISIT_OPTION_DEFAULT(VISIT_HDF4_DIR     ${VISITHOME}/hdf/4.2.13)
       JPEG_LIBRARY_DIR JPEG_LIB
       SZIP_LIBRARY_DIR SZIP_LIB
-      ZLIB_LIBRARY_DIR ZLIB_LIBRARY
+      ZLIB_LIBRARY_DIR ZLIB_LIB
```
Am confused...in preceding CMakeLists.txt files `ZLIB_LIB` was replaced with `ZLIB_LIBRARY`. Here, the reverse appears to be happening.

biagas replied:
our dependency checking requires listing of the library location, then the library, separately.  When adding libraries to target_link_libraries, cmake prefers full path to library.  ZLIB_LIB is just the library name, ZLIB_LIBRARY is the full path.

markcmiller86 commented on src/databases/ADIOS2/ADIOS2.xml:
```diff
@@ -11,6 +11,7 @@
       avtADIOS2FileFormat.C
       avtMEUMMAPSFileFormat.C
       avtGTCFileFormat.C
+      avtLAMMPSFileFormat.C
```
Is this related to zlib changes? If not, thats ok. Am not suggesting you make a new branch/commit/PR.

markcmiller86 commented on src/databases/BOV/avtBOVFileFormat.C:
```diff
@@ -42,7 +42,7 @@
 #include <avtBOVFileFormat.h>
 
-#include <vtk_zlib.h>
+#include <zlib.h>
```
Shouldn't these be conditioned like so...

```
#ifdef HAVE_ZLIB_H
#include <zlib.h>
#endif

biagas replied:
What about the other sections of the plugin code that require zlib. Should they all be conditioned, or should the entire plugin be conditioned on the presence of zlib in the databases/CMakeLists.txt?

markcmiller86 replied:
Depends. Does the plugin not function *at all* without zlib or does the plugin just loose some features? If the former, disable plugin entirely. Otherwise, condition zlib.h inclusion.

biagas replied:
I think BOV plugin in its entirety should depend on zlib being found. It can open files with '.gz' extension. Unless we add a warning to the plugin when user attempts to open a gzipped file and we don't have zlib support.  Plus there are a couple sections in the writer where I'm not certain the file will be written properly without zlib. See section of writer starting at line 719.

markcmiller86 replied:
When you say it *can* open files with `.gz` extension, do you mean it indeed *requires* them to be `.gz` (gzip compressed) files and there is no coding to allow it to read data from UNcompressed files. If so, I agree that when zlib is NOT enabled, the BOV plugin should also be NOT enabled.

biagas replied:
I am working through the code that uses gz. Easy enough to wrap it with preprocessor directives if we don't have zlib.  However, I am not familiar enough with databases to know when to bail out and issue a warning/error if BOV should be able to read a gz file but cannot because zlib wasn't found. It looks like only the GetVar method looks for the extension.

markcmiller86 replied:
Ok, so the use case is...a user goes to open a compressed BOV file and VisIt fails to open it because the VisIt they are using was NOT compiled with zlib support. Is that right? And, you'd like to issue a descriptive error message instead of silently failing? 

I think the only way can detect that condition in the plugin is if you read the first few bytes of the file and they match (one of) the zlib header magic number, and this is not a zlib build, you can issue an error message. Maybe more trouble than its worth though.

biagas replied:
Yes, that was my thought, give the user some indication as to why they cannot open their compressed BOV file.

markcmiller86 replied:
Ok, here's what I would do. Around [this line](https://github.com/visit-dav/visit/blob/37262d6cf642a3b0f0b29b11919989389c00ca6e/src/databases/BOV/avtBOVFileFormat.C#L1188), add logic similar to following

```
 size_t nread = fread(rv->GetVoidPointer(0), unit_size,
                                 nvals * dataNumComponents, file_handle);

if (nread != unit_size * nvals * dataNumComponents) // fread failed in some way
{
    unsigned short magic = 0x0000;
    fseek(file_handle, 0, SEEK_SET);
    fread(&magic, 2, 1, file_handle);
    if (magic == 0x1F8B || magic == 0x8B1F) // avert endieness issues by checking both orders
    {
        // Probable attempt to read compressed BOV data with a plugin that was not compiled with zlib enabled.
    }
}
```

There are various places gzip magic number is described. [Here's one](https://en.wikipedia.org/wiki/List_of_file_signatures) -- search for `gzip`.

There are two places in the file to put this check so maybe create a local `avtBOVFileFormat::Read()` method to encapsulate it.

markcmiller86 replied:
oops, you need a `fseek()` in there too to rewind the file back to beginning *before* doing the `fread` for magic number.

biagas replied:
Isn't the GetVar method a little late in the process to be bailing on the reader?

markcmiller86 replied:
I don't think so. Can you identify an earlier point? Note...BOV header is uncompressed ascii. There is no point in compressing this small file. If BOV data is compressed, only the variable data will be. The reader can become aware of a problem related to compression *only* after its made an attempt to read a compressed file. And, we're a bit lucky in this regard because each variable is in its *own* file so a *short* read is detectable. If multiple variables were in the same file, a *short* read would not be detectable because the read would read past the first compressed variable and into the others and we'd be none the wizer. Come to think of it, this *is* a problem even for a single variable per file because some variable's *compressed* file may in fact wind up being *larger* than the uncompressed data. We cannot detect a *short* read in that case either. That would mean VisIt would read the data, thinking is it fine and all hell would break loose (probably with a lot of NaNs) later with no clue as to why.

Without detecting a read failure, the only option we have left is to inspect the first two bytes of the variable file for tell-tail signs its gzip'd (gzip magic number), which I think is potentially problematic because we could have real float/double binary data at the zero offset in the file that would match that, I think.

So, I think it *has* to be the combination of a read failure *and* that magic before we can reasonably assume its compressed data.

I do not see anything in the BOV header that can be used to explicitly indicate if data is compressed or not (even though our Getting Data Into VisIt Manual in fact mentions that as an option). However, since BOV appears to be a format we/VisIt created, you could decide to add such a keyword there. But, that would effect only new BOV data created going forward and not any existing BOV data.

Another route to avoid all these issues entirely is to just decide to always build VisIt with zlib support and I think that would be a fine idea.

biagas replied:
So, make zlib a required library in build_visit, avoiding use of system zlib?
Or, still allow build_visit to use either system zlib or one built with build_visit.

For VisIt itself, make zlib a required library such that if FindZlib fails to find either a system or built version of zlib, configure cannot continue?

Or, just make the plugins in question not be built if we don't have zlib?  

I think in most cases (for most platforms?), we will find zlib, either the one built with build_visit or a system version.

markcmiller86 replied:
Not sure. AFAICT, only risk in *requiring* zlib is chance that it won't build somewhere we need it. I think it is highly unlikely that zlib would NOT build where we need it but rest of VisIt would build...which would mean zlib is killing our ability to build somewhere.

Since zlib is so ubiquitous, in 99% of the cases, *requiring* it would mean just adding `-lz` to `LDFLAGS`.

Just another avenue to consider and which might simplify many of these issues.

markcmiller86 commented on src/databases/OpenFOAM/visit_vtkOpenFOAMReader.cxx:
```diff
@@ -59,7 +59,7 @@
 #include <vtksys/SystemTools.hxx>
 #include <sstream>
-#include <vtk_zlib.h>
+#include <zlib.h>
```
Shouldn't this be conditioned on `HAVE_ZLIB_H`?

biagas replied:
I believe this reader should be conditioned on having zlib. I think it would be too difficult to wrap the code sections that deal with compressed files.

markcmiller86 replied:
When I look at at [visit_vtkOpenFoamReader.cxx](https://github.com/visit-dav/visit/blob/684e82f2e5a18f4f9495d49c3c066c70c8cf6059/src/databases/OpenFOAM/visit_vtkOpenFOAMReader.cxx#L1447), I see logic suggestive that it is capable of reading either compressed or UNcompressed files. On that simple check, my guess is that it should not get disabled with ZLIB is not present and instead, `#Iinclude <zlib.h>` should be conditined on `#HAVE_ZLIB_H`

biagas replied:
As I stated before, I think the logic in this file is too complex to wrap all the gz stuff with preprocessor checks.

markcmiller86 replied:
Ok, so I guess I was misunderstanding...you are saying that it is NOT a simple matter of including or not including `zlib.h` to support compressed inputs here because there are also many other places that the file would need conditional compilation blocks

markcmiller86 commented on src/tools/dev/scripts/bv_support/bv_hdf4.sh:
```diff
@@ -16,7 +16,11 @@ function bv_hdf4_disable
+    if [[ "$DO_ZLIB" == "yes" ]] ; then
+        echo "szip zlib"
+    else
+        echo "szip"
```
So, HDF4 depends UNCONDITIONALLY on szip?

biagas replied:
HDF4 previously unconditionally depended upon both szip and zlib being built by build_visit.  I simply removed zlib from being required to be built by build_visit. If zlib not requested, system zlib is attempted.

markcmiller86 commented on src/tools/dev/scripts/bv_support/bv_zlib.sh:
```diff
@@ -57,6 +57,14 @@ function bv_zlib_host_profile
 
+function bv_zlib_initialize_vars
+{
+    export VISIT_ZLIB_DIR=${VISIT_ZLIB_DIR:-"$VISITDIR/zlib/${ZLIB_VERSION}/${VISITARCH}"}
```
This is for the case where zlib is supported via an explicit build of it. What about the case where system zlib is used? What gets exporte for VISIT_ZLIB_DIR?

biagas replied:
Nothing should be exported, I will make the export conditional.

markcmiller86 replied:
> I will make the export conditional.

Did you alread commit that?

biagas replied:
Yes.  Did you notice the 'Outdated' label above.  You can see the changes made in the 'Files changed' view.

markcmiller86 replied:
Yes, I did. I followed the link to the file and then selected from `...` in upper right, `view file` and did not see it. My mistake however as `view file` from that screen takes me to the specific commit hash on that file and not the head of the branch. I just looked at the head and see what I expect there now. Sorry.

biagas committed:
Make export of zlib location dependent upon zlib being requested in the build.

markcmiller86 commented:
AFICT, there are several unresolved/unanswered questions still...So, I am waiting for those to be addressed.

biagas committed:
Make gz logic dependent upon HAVE_ZLIB_H being defined.

biagas committed:
Add comment.

biagas commented:
@markcmiller86   Could you please Resolve conversations when you feel they have been adequately addressed?

markcmiller86 commented:
> @markcmiller86 Could you please Resolve conversations when you feel they have been adequately addressed?

Ok, I've done that for the ones I though were resolved. I honestly don't know if I've missed something you are waiting on from me though.

biagas committed:
Wrap gz code in #ifdef HAVE_ZLIB_H.

markcmiller86 reviewed (changes requested):
Something seems a miss here because these are all `#if 0` which will *never* compile under any condition.

biagas committed:
Fix #if 0 which were for testing compilation without zlib.

biagas commented:
Sorry about the #if 0. Needed to switch to a different branch in a hurry, and just committed the file. The #if 0 was there for testing compilation when zlib not present. Fixed. now.

biagas committed:
Make zlib (either system or one built with build_visit) required to build visit.
Remove HAVE_ZLIB_H coding from databases requiring zlib.
Fix some whitespace issues.

biagas committed:
Merge branch 'develop' into bugfix/biagas/consistent_zlib_usage

biagas committed:
Make zlib a required library in build_visit.
Fix a few build modules to reflect that zlib should always be built.

biagas committed:
Merge branch 'bugfix/biagas/consistent_zlib_usage' of https://github.com/visit-dav/visit into bugfix/biagas/consistent_zlib_usage

biagas committed:
Make zlib enabled by default.
